lib/maze/docker.rb

```diff
@@ -0,0 +1,108 @@
+      # If running a command, it will be executed as an attached process, otherwise it
+      # will run detached.
+      #
+      # @param service [String] The name of the service to start
+      # @param command [String] Optional. The command to use when running the service
+      # @param interactive [Boolean] Optional. Whether to run interactively
+      def start_service(service, command: nil, interactive: false)
+        if interactive
+          run_docker_compose_command("build #{service}")
+
+          # Run the built service in an interactive session. The service _must_
+          # have an appropriate entrypoint, e.g. '/bin/sh'. We also disable ANSI
+          # escape sequences from docker-compose as they can cause issues with
+          # stderr expectations by 'leaking' into the next line
+          command = get_docker_compose_command("--no-ansi run #{service} #{command}")
+
+          cli = Runner.start_interactive_session(command)
+          cli.on_exit do |status|
+            @last_exit_code = status
+            @last_command_logs = cli.stdout_lines + cli.stderr_lines
+          end
+
+          # The logs and exit code aren't available from the interactive session
+          # at this point (we've just started it!) so we can't provide them here
+          @last_command_logs = []
+          @last_exit_code = nil
+        elsif command
+          # We build the service before running it as there is no --build
+          # option for run.
+          run_docker_compose_command("build #{service}")
+          run_docker_compose_command("run --use-aliases #{service} #{command}")
+        else
+          # TODO: Consider adding a logs command here
+          run_docker_compose_command("up -d --build #{service}")
+        end
+        @services_started = true
+      end
+
+      # Kills a running service
+      #
+      # @param service [String] The name of the service to kill
+      def down_service(service)
+        # We set timeout to 0 so this kills the services rather than stopping them
+        # as its quicker and they are stateless anyway.
+        run_docker_compose_command("down -t 0 #{service}")
+      end
+
+      # Resets any state ready for the next scenario
+      def reset
+        down_all_services
+        @last_exit_code = nil
+        @last_command_logs = nil
+      end
+
+      # Kills all running services
+      def down_all_services
+        # This will fail to remove the network that maze is connected to
+        # as it is still in use, that is ok to ignore so we pass success codes!
+        # We set timeout to 0 so this kills the services rather than stopping them
+        # as its quicker and they are stateless anyway.
+        run_docker_compose_command('down -t 0', success_codes: [0, 256]) if compose_stack_exists? && @services_started
+        @services_started = false
+      end
+
+      def compose_project_name
+        @compose_project_name ||= nil
+      end
+
+      attr_writer :compose_project_name
```

Comment:
I don't fully understand what's going on, but I _think_ it's needed because the class uses `class << self`.  If I remove the setter and change `compose_project_name` to `@compose_project_name` on L92 it falls over in a heap.